### PR TITLE
[IaC] feat: fetch policy bundle from S3 for local scans [Beta]- CC-627

### PR DESCRIPTION
**Pull request by ipapast**
#### What does this PR do?
_This is the basic e2e flow that we need for the Beta release. It will be incrementally enhanced._
Everytime a user runs `snyk iac --experimental`, the wasm policies will be available locally.
What happens in the background: 
- A tarball bundle.tar.gz is downloaded from S3, is extracted and placed in the `.iac-data` directory of your cwd. 

This way, the user will have the policies locally in a wasm file, giving them the ability to scan their files locally against them and not send any files over the network.

**Regarding the smoke tests and the env variable:** 
these are run against a subset of fixtures, but now that we introduce this new flow, the policies would get updated every time. Because of that, we provided an env variable temporarily named `SNYK_IAC_SKIP_BUNDLE_DOWNLOAD` to shellspec to skip downloading the bundle.

#### How should this be manually tested?
- Run `snyk iac test --experimental` as usual to check the full flow.
- Run `SNYK_IAC_SKIP_BUNDLE_DOWNLOAD=true snyk-dev iac test filename --experimental` to skip the download of the bundle from S3, for that you will need to have the `.iac-data` folder with the relevant files locally, or you will get an error.

#### Any background context you want to provide?
There is [this decision document](https://www.notion.so/snyk/Policy-WASM-download-mechanism-8b25b39dceab4520b56f9783e6cd7ecf) in Notion that can be shared internally.

#### What are the relevant tickets?
[CC-627](https://snyksec.atlassian.net/browse/CC-627)

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -7,6 +7,9 @@ module.exports = {
   coverageReporters: ['text-summary', 'html'],
   testMatch: [
     '<rootDir>/test/*.spec.ts',
+    '<rootDir>/test/iac-unit-tests/*.spec.ts',
+    '<rootDir>\\test\\*.spec.ts', // for Windows,
+    '<rootDir>\\test\\iac-unit-tests\\*.spec.ts', // for Windows
```
Comment (rontalx) [inline jest.config.js:12]: should be able to rebase from master and have these now.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -14,6 +14,7 @@ test/acceptance/cli-test/iac/ @snyk/cloudconfig
 test/fixtures/iac/ @snyk/cloudconfig
 test/smoke/spec/iac/ @snyk/cloudconfig
 test/smoke/.iac-data/ @snyk/cloudconfig
+test/iac-unit-tests/ @snyk/cloudconfig
```
Comment (rontalx) [inline .github/CODEOWNERS:17]: should be able to rebase from master and have these now.

File: src/lib/types.ts [hunk 3]
```diff
@@ -78,6 +78,7 @@ export interface Options {
   debug?: boolean;
   sarif?: boolean;
   'group-issues'?: boolean;
+  cacheDir?: boolean;
```
Comment (rontalx) [inline src/lib/types.ts:81]: nit:
got a comment from Ben to rename this to: `cache-directory`

File: src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import * as fs from 'fs';
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts:1]: I thought of bundling all of this code inside the `local-cache.ts` module, so it handles all logic related to the local cache and downloading.
Maybe we can rename it to something like `policy-provider.ts` or so.
WDYT?

File: src/cli/commands/test/iac-local-execution/index.ts [hunk 5]
```diff
@@ -1,21 +1,31 @@
...
+    await fetchBundle(preSignedUrl);
+  }
+  if (!isLocalCacheExists()) {
     throw Error(
       `Missing IaC local cache data, please validate you have: \n${REQUIRED_LOCAL_CACHE_FILES.join(
         '\n',
       )}`,
     );
+  }
+
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/index.ts:28]: Following:
https://github.com/snyk/snyk/pull/1690/files#r589434889
Can we move these lines into the unified module?
so all cache/download logic is encapsulated there.
Then also the test command will keep a cool neat convention, for example something like that:
```
  import { initPolicyProvider } from './local-cache.ts'
  ....
  await initPolicyProvider(options);     //
  const filesToParse = await loadFiles(pathToScan);
  const { parsedFiles, failedFiles } = await parseFiles(filesToParse);
  const scannedFiles = await scanFiles(parsedFiles);
  const formattedResults = formatScanResults(scannedFiles, options);
```

File: src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import { Readable } from 'stream';
+
+export async function fetchBundle(preSignedUrl: string) {
+  createIacDir();
+  const bundlePath: fs.PathLike = path.join('.iac-data/bundle.tar.gz');
+
+  https.get(preSignedUrl, (response) => {
+    streamToFile(response, bundlePath);
+  });
+  return extractBundle(bundlePath);
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts:14]: Although this approach with is cool, I'd prefer to avoid this stream & callback implementation.
it's hard to comprehend from first look, and I don't think it adds value due to the synchronous nature of this flow.
I'd suggest checking if you can promisify the `https.get` (I know we use `needle` in the CLI so maybe that can be utilised).
so the flow eventually is as simple as:
```
// await on downloading the file
// synchronously unzipping it with tar
// synchronously write it to the directory.
```
it's easier to read and the readers doesn't need to wrap their heads around asynchronous streams/pipes and callbacks, just to realise it's a simple synchronous flow. 
on other scenarios it can be great, I think its less convenient here.
WDYT?

File: src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  const iacPath: fs.PathLike = path.join('.iac-data/');
+  try {
+    if (!fs.existsSync(iacPath)) {
+      fs.mkdirSync(iacPath, '700');
+    }
+    fs.accessSync(iacPath, fs.constants.W_OK);
+  } catch {
+    throw Error(
+      'The .iac-data directory can not be created or written. ' +
+        'Please check that the current directory has write permissions',
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts:27]: Looking good, I'd double check with Ben for the specific wording.

File: test/iac-unit-tests/fetch-wasm-bundle.spec.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { streamToFile } from '../../src/cli/commands/test/iac-local-execution/fetch-wasm-bundle';
```
Comment (rontalx) [inline test/iac-unit-tests/fetch-wasm-bundle.spec.ts:1]: you mentioned that you're adding tests, 
double checking you're planning to test the public/exposable method of the module `fetchBundle`?

File: src/cli/commands/test/iac-local-execution/index.ts [hunk 9]
```diff
@@ -1,21 +1,31 @@
...
+import { fetchBundle } from './fetch-wasm-bundle';
 
 // this method executes the local processing engine and then formats the results to adapt with the CLI output.
 // the current version is dependent on files to be present locally which are not part of the source code.
 // without these files this method would fail.
 // if you're interested in trying out the experimental local execution model for IaC scanning, please reach-out.
+
 export async function test(pathToScan: string, options) {
-  if (!isLocalCacheExists())
+  if (!options.cacheDir) {
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/index.ts:15]: @benlaplanche @ipapast 
Following our sync today, double checking we're aligned...
For the beta-release:
- the `--cache-directory` can be provided, which will avoid the forced downloading.
- in this release the `--cache-directory` serves merely as a boolean flag to allow avoid the forced download every time. emphasising it is still not supportive of providing a custom directory, this will be added for the GA. 
- open question: how would we expect the `--cache-directory` flag to work if the bundle wasn't downloaded yet, download it? or error out and expect a first iteration without it to run in order to do the initial bundle fetch.
I'd say, `--cache-directory` should download a bundle if there's non present (will happen for the first run with that flag.).

Although an edge-case, I think it can be impactful in terms of experience.
WDYT?

**Comment by rontalx** [comment]
great work! left some comments.

File: src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts (hunk 6 above)
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/fetch-wasm-bundle.ts]: Thanks, I changed this part to be synchronous as you suggested, I hope it reads better now

File: src/cli/commands/test/iac-local-execution/file-utils.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import * as fs from 'fs';
+import { WriteStream } from 'fs';
+import * as tar from 'tar';
+import * as path from 'path';
+import ReadableStream = NodeJS.ReadableStream;
+
+export function createIacDir(): void {
+  const iacPath: fs.PathLike = path.join('.iac-data/');
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/file-utils.ts:8]: mind adding small `TODO` notes on the places it uses `'.iac-data/'` that in the future it would be custom?

File: src/cli/commands/test/iac-local-execution/index.ts (hunk 9 above)
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/index.ts]: reverting this flag as discussed

Comment (aron) [reply src/cli/commands/test/iac-local-execution/index.ts]: Not for now, but my understanding of this flag, in the GA state based on experience with other tools is as follows.

1) Without the flag Snyk will default to `$(pwd)/.snyk-iac`.
  1) If the .snyk-iac directory exists the CLI will check if the contained policies are present, valid and up-to-date. And if stale either prompt to download or download automatically in the background.
2) With the `--cache-directory` flag the CLI will replace the `$(pwd)/.snyk-iac` path with that provided by the user. The same steps as above apply.

I don't think the flag should have any impact on the download behaviour, it should simply be used to configure a path.

Comment (aron) [reply src/cli/commands/test/iac-local-execution/index.ts]: > instead of just running with the flag, and have it download the files to the provided directory if it's empty.

This is what I intended to say. I've updated the comment to hopefully clarify.

Comment (rontalx) [reply src/cli/commands/test/iac-local-execution/index.ts]: so we are talking on the same thing 🤓 
Anyway for now let's extract this flag from this PR for simplicity.
We can add it in the coming sprint.

File: src/cli/commands/test/iac-local-execution/file-utils.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    inputStream
+      .pipe(fileWriteStream)
+      .on('finish', resolve)
+      .on('error', reject);
+  });
+}
+
+export function extractBundle(bundlePath: fs.PathLike): Promise<void> {
+  return new Promise((resolve, reject) => {
+    fs.createReadStream(bundlePath)
```
Comment (aron) [inline src/cli/commands/test/iac-local-execution/file-utils.ts:37]: Not for now, but why write the bundle to disk when we can pipe the HTTP stream directly into `tar.x()`?

File: .circleci/config.yml [hunk 12]
```diff
@@ -207,7 +207,7 @@ jobs:
             echo "Checkout the README in test/smoke folder for more details about this step"
             unset SNYK_API
             unset SNYK_API_KEY
-            shellspec -f d -e REGRESSION_TEST=1
+            shellspec -f d -e REGRESSION_TEST=1 -e SNYK_IAC_SKIP_BUNDLE_DOWNLOAD=true
```
Comment (ipapast) [inline .circleci/config.yml:210]: This is to provide an env var for IaC smoke tests -we use local fixtures in the tests but now the implementation downloads new policies everytime. We wanted to skip the download for now by adding this var.

File: src/cli/commands/test/iac-local-execution/file-utils.ts (hunk 11 above)
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-utils.ts:37]: changed this, writing is removed now

File: src/cli/commands/test/iac-local-execution/index.ts [hunk 13]
```diff
@@ -1,21 +1,17 @@
...
 // if you're interested in trying out the experimental local execution model for IaC scanning, please reach-out.
+
 export async function test(pathToScan: string, options) {
-  if (!isLocalCacheExists())
-    throw Error(
-      `Missing IaC local cache data, please validate you have: \n${REQUIRED_LOCAL_CACHE_FILES.join(
-        '\n',
-      )}`,
-    );
+  await initLocalCache();
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/index.ts:14]: yasss ✨

File: src/cli/commands/test/iac-local-execution/local-cache.ts [hunk 14]
```diff
@@ -46,3 +49,23 @@ export function getLocalCachePath(engineType: EngineType) {
       ];
   }
 }
+
+export async function initLocalCache(): Promise<void> {
+  // temporarily use an ENV var to skip downloading of the bundles if we need to - e.g. smoke tests
+  if (!process.env.SNYK_IAC_SKIP_BUNDLE_DOWNLOAD) {
+    const preSignedUrl =
+      'https://cloud-config-policy-bundles.s3-eu-west-1.amazonaws.com/bundle.tar.gz';
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/local-cache.ts:57]: nit:
maybe worth adding a small TODO/comment that it will be replaced with snyk . io domain soon

**Review by rontalx — APPROVED** [review]
great stuff 👏

File: jest.config.js (hunk 1 above)
Comment (ipapast) [reply jest.config.js:12]: Those were there already. Line 13 has an extra comma in the end of the comment and didn't cause conflicts. I can remove the line.

File: src/cli/commands/test/iac-local-execution/local-cache.ts (hunk 14 above)
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/local-cache.ts:56]: We haven't taken a decision officially with what it would be replaced exactly so I think I'll skip this

File: src/cli/commands/test/iac-local-execution/local-cache.ts [hunk 15]
```diff
@@ -46,3 +49,23 @@ export function getLocalCachePath(engineType: EngineType) {
       ];
   }
 }
+
+export async function initLocalCache(): Promise<void> {
+  // temporarily use an ENV var to skip downloading of the bundles if we need to - e.g. smoke tests
+  if (!process.env.SNYK_IAC_SKIP_BUNDLE_DOWNLOAD) {
```
[the remark this answers is not in the document]
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/local-cache.ts:54]: Just personal preference I guess 😅 I thought it would be much easier to clean afterwards

File: src/cli/commands/test/iac-local-execution/file-utils.ts (hunk 10 above)
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-utils.ts:6]: done

File: jest.config.js [hunk 16]
```diff
@@ -7,10 +7,12 @@ module.exports = {
   coverageReporters: ['text-summary', 'html'],
   testMatch: [
     '<rootDir>/test/*.spec.ts',
-    '<rootDir>\\test\\*.spec.ts', // for Windows
     '<rootDir>/test/iac-unit-tests/*.spec.ts',
     '<rootDir>\\test\\*.spec.ts', // for Windows,
     '<rootDir>\\test\\iac-unit-tests\\*.spec.ts', // for Windows
+    '<rootDir>\\test\\*.spec.ts', // for Windows
+    '<rootDir>/test/iac-unit-tests/*.spec.ts',
+    '<rootDir>\\test\\iac-unit-tests\\*.spec.ts', // for Windows
```
Comment (jahed-snyk) [inline jest.config.js:15]: It looks like Line 10 has been moving to Line 13 and Line 14 and 15 are duplicates of Line 11 and 13. Rolling back all changes to `jest.config.js` in this PR should fix this.

File: .circleci/config.yml (hunk 12 above)
Comment (jahed-snyk) [reply .circleci/config.yml:210]: I'm not sure about this change. The goal of smoke tests is to capture production issues so skipping a production step kind of defeats its purpose. Mentioning @JackuB as I don't have full context of smoke tests.

**Review by jahed-snyk — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 12 above)
Comment (ipapast) [reply .circleci/config.yml:210]: Hi @jahed-snyk , @JackuB, I just want to clarify that we are not skipping the tests but the download of an artifact. The download step now overwrites the local fixtures we are using in the smoke tests on every download, so the expected results change. 
We introduced this var temporarily, we will think of how to do this better and according to the next steps of this flow, but for now we want the tests to run against the fixed set of fixtures.

Does this make more sense?

Comment (jahed-snyk) [reply .circleci/config.yml:210]: Thanks for clarifying. That makes sense.

Bear with me for a bit as I get some feedback on the rest of the changes from the team (I don't have enough context).

File: src/cli/commands/test/iac-local-execution/local-cache.ts [hunk 17]
```diff
@@ -46,3 +48,23 @@ export function getLocalCachePath(engineType: EngineType) {
...
+    const preSignedUrl =
+      'https://cloud-config-policy-bundles.s3-eu-west-1.amazonaws.com/bundle.tar.gz';
```
Comment (rontalx) [inline src/cli/commands/test/iac-local-execution/local-cache.ts:56]: following our conversation and drift of the bundle,
let's make sure to use here an explicit bundle name with a tagged version so we keep track of what is being used., and we know is working.
as we're not covering the download scenario of specific bundles in tests yet.
❗️🚨

Comment (rontalx) [reply src/cli/commands/test/iac-local-execution/local-cache.ts:56]: ok after what we discussed now, ignore this :)
